### PR TITLE
Implement the extended scope and further improvements for ktxtools

### PR DESCRIPTION
**Pull request by VaderY**
This PR completes the original and the extended scope of ktxtools. Some of the scope was already part of the previous PRs.

Briefs:
- tools: Implement stdout support
- tools: Implement stdin support
- tools: Implement 4:2:2 support
- tools: Implement support for 10X6 and 10X4 formats
- tools: Implement support for B10G11R11_UFLOAT and E5B9G9R9_UFLOAT formats
- tools: Complete support for Depth-Stencil formats
- tools: Improvements, cleanup and bug fixes
- extract: Implement fragment URI support
- extract: Implement 4:2:2 sub-sampling
- validate: Fix and extend padding byte validation checks
- cts: Add support for stdin/stdout test cases (including binary IO)
- cts: Add tests to cover new features and capabilities
- cts: Extend existing tests to improve coverage
- cts: Test runner now deletes matching output files (this enables easy packaging of mismatching files)
- cts: Added a new cli arg to the runner script: --keep-matching-outputs to prevent the deletion of matching output files
- dfdUtils: Implement 4:2:2 support
- dfdUtils: Implement support for 10X6 and 10X4 formats
- imageio: Fix stdin support
- ktx: Add stringToVkFormat to mkvkformatfiles
- ktx: Implement 3D BC support (ASTC 3D Blocks)
- ktx: Implement 4:2:2 support
- ktx: Complete support for Depth-Stencil formats
- ktx: Improve interpretDFD
- ktx: Improvements, cleanup and bug fixes
- cmake: Add CMake generator expression for output directory on Mac

Please complete the final review of this scope including the CTS.

### CODE REVIEW

File: tools/ktx/command.h [hunk 1]
```diff
@@ -289,23 +294,34 @@ struct OptionsMultiInSingleOut {
...
         if (files.size() < 2)
-            report.fatal_usage("Output file must be specified.");
+            report.fatal_usage("{} file must be specified.", args.count("stdout") == 0 ? "Output" : "Input");
 
         outputFilepath = std::move(files.back());
         files.pop_back();
         inputFilepaths = std::move(files);
+
+        if (std::count(inputFilepaths.begin(), inputFilepaths.end(), "-") > 1)
+            report.fatal_usage("'-' or --stdin as input file was specified more than once.");
```
Comment (MarkCallow) [inline tools/ktx/command.h:324]: Is the CLI syntax to write `--stdin` in the input file list where you want stdin used, e.g. `a --stdin c`?

File: tools/ktx/command_extract.cpp [hunk 2]
```diff
@@ -296,14 +235,57 @@ void CommandExtract::OptionsExtract::process(cxxopts::Options&, cxxopts::ParseRe
...
+        report.fatal_usage("stdout cannot be used with multi-output '--depth all' extract.");
+
+    if (args["uri"].count()) {
+        uriFlagUsed = true;
+        try {
+            fragmentURI = parseFragmentURI(args["uri"].as<std::string>());
+        } catch (const std::exception& e) {
+            report.fatal_usage("Failed to parse Fragment URI: {}", e.what());
+        }
+        if (globalAll)
```
Comment (MarkCallow) [inline tools/ktx/command_extract.cpp:265]: Why not do these ifs, up to line 272,  before parsing the fragment URI?

**Review by MarkCallow — APPROVED** [review]
Looks good. I have several questions. A couple may result in you making code changes.

While reviewing this I spotted another feature of `toktx` that has not been brought forward to the new tools.

    infiles prefixed with '@' are read as text files
    listing actual file names to process with one file path per line. Paths
    must be absolute or relative to the current directory when toktx is run.
    If '@@' is used instead, paths must be absolute or relative to the location
    of the list file.

Let's see if we get complaints from early users. When I added this to toktx it was not very difficult so I may try to add it myself, at some point in the future, with tests of course.

File: tools/ktx/CMakeLists.txt [hunk 3]
```diff
@@ -43,7 +44,14 @@ PRIVATE
     .
     $<TARGET_PROPERTY:imageio,INCLUDE_DIRECTORIES>
     $<TARGET_PROPERTY:ktx,INCLUDE_DIRECTORIES>
+)
+
+target_include_directories(
```
[the remark this answers is not in the document]
Comment (VaderY) [reply tools/ktx/CMakeLists.txt:49]: `SYSTEM` is a flag that effects every item in `target_include_directories` so these must be two separate call.

File: tests/unittests/test_fragment_uri.cc [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+// #include "gl_format.h"
+// #include "ktx.h"
+// extern "C" {
+//   #include "ktxint.h"
+//   #include "filestream.h"
+//   #include "memstream.h"
+// }
+#include "gtest/gtest.h"
+// #include "wthelper.h"
+// #include "ltexceptions.h"
```
[the remark this answers is not in the document]
Comment (VaderY) [reply tests/unittests/test_fragment_uri.cc:22]: They were just leftover, I cleaned them up.

File: tools/imageio/imageinput.cc [hunk 5]
```diff
@@ -83,7 +83,7 @@ ImageInput::open(const _tstring& filename,
             buffer =
                 std::unique_ptr<std::stringstream>(new std::stringstream);
             *buffer << std::cin.rdbuf();
-            buffer->seekg(0, buffer->beg);
+            buffer->seekg(0, std::ios::beg);
```
[the remark this answers is not in the document]
Comment (MarkCallow) [reply tools/imageio/imageinput.cc:86]: No problem. I had a hard time remembering which class `beg` and `end` are defined in so started referencing via the object.

File: tools/ktx/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarkCallow) [reply tools/ktx/CMakeLists.txt:49]: Thanks for clarification.

File: tools/ktx/command.cpp [hunk 6]
```diff
@@ -43,4 +46,97 @@ std::string version(bool testrun) {
...
+        // std::freopen(nullptr, "wb", stdout);
+        #endif
+        file = stdout;
+    } else {
+        file = std::fopen(filepath.c_str(), "wb");
+        if (!file)
+            report.fatal(rc::IO_FAILURE, "Could not open output file \"{}\": {}.", filepath, errnoMessage());
+    }
+
+    // if (filepath == "-") {
```
[the remark this answers is not in the document]
Comment (VaderY) [reply tools/ktx/command.cpp:93]: I had portability issues with the C++ version. I think the cause was `StreambufStream` but I did not investigate it fully.

[the remark this answers is not in the document]
Comment (MarkCallow) [reply tools/ktx/command.cpp:93]: Please add a comment to that effect for future reference.

File: tools/ktx/command.h (hunk 1 above)
Comment (VaderY) [reply tools/ktx/command.h:324]: `--stdin` will prepend a single `-` to the input file list so if there are multiple input files than the first file will be read from the stdin.
`-` is supported both as input and output for aliasing stdin and stdout and its position can vary freely in the input/output list

File: tools/ktx/command_extract.cpp (hunk 2 above)
Comment (VaderY) [reply tools/ktx/command_extract.cpp:265]: Yeah, It would make more sense. I have reordered them.

File: tools/ktx/command.h (hunk 1 above)
Comment (MarkCallow) [reply tools/ktx/command.h:324]: That wasn't entirely clear to me from the option help lines that I read while reviewing. Please improve the option help lines in `command.h`. (The line here is fine. It is just where I realized the possible confusion.)

File: tools/ktx/command.h [hunk 7]
```diff
@@ -289,23 +294,35 @@ struct OptionsMultiInSingleOut {
 
     void init(cxxopts::Options& opts) {
         opts.add_options()
-                ("files", "Input/output files. Last file specified will be used as output", cxxopts::value<std::vector<std::string>>(), "<filepath>");
+                ("stdin", "Use stdin as the first input file. (Using a single dash '-' as the first input file has the same effect)")
+                ("stdout", "Use stdout as the output file. (Using a single dash '-' as the output file has the same effect)")
+                ("files", "Input/output files. Last file specified will be used as output."
+                          " Using a single dash '-' as the an input or output file will use stdin/stdout.", cxxopts::value<std::vector<std::string>>(), "<filepath>");
```
Comment (MarkCallow) [inline tools/ktx/command.h:300]: s/the an/an/

The rest of the help changes look good.

**Pull request merged**